taichi/codegen/spirv/spirv_codegen.cpp

```diff
@@ -1647,12 +1682,18 @@ class TaskCodegen : public IRVisitor {
       return it->second;
     }
 
-    if (buffer.type == BufferType::Context) {
+    if (buffer.type == BufferType::Args) {
       buffer_binding_map_[key] = 0;
       buffer_value_map_[key] = args_buffer_value_;
       return args_buffer_value_;
     }
 
+    if (buffer.type == BufferType::Rets) {
+      buffer_binding_map_[key] = 1;
+      buffer_value_map_[key] = ret_buffer_value_;
+      return ret_buffer_value_;
+    }
+
     int binding = binding_head_++;
     buffer_binding_map_[key] = binding;
```

Comment:
I don't quite understand whats happening here. If we return early for "Args" and "Rets", doesn't that mean `buffer_argument` won't be called for these two buffers? E.g. they would miss the "Block" decoration. This was also the case in #4205, and I've been getting complaints from tint because of that.

Also, wouldn't this cause Args and Rets to be bound to the same binding point as other buffers?

Reply:
Or maybe the "Block" decoration should be added to `buffer_struct_argument`?

Reply:
Another thing about #4205 and this PR that's a bit problematic:  In 4205, `compile_args_struct()` gets called regardless of whether or not there're any arguments. You fixed that in this PR by returning immediately if num args = 0, but not every offloaded task in the kernel is guaranteed to use those args. So you could end with a offloadedTask/shader that declares the args buffer as an input, but never use it.  This might be fine with Vulkan (i'm not sure), but it's causing a lot of trouble for webgpu stuff.

Reply:
Let's call "compile_args" only when we notice a `ArgLoadStmt`. Same for return values.

Reply:
The decoration is added when calling "compile_args_struct" where it calls the “buffer_struct_argument" or "uniform_struct_argument"

Reply:
Also the `binding_head_` is initialized to 2 so other things won't bind on 0 and 1. It's quite ugly, but it's the simplest solution I have

Reply:
> binding_head_ is initialized to 2 

Right. That works. But please add a comment tho.

> The decoration is added when calling "compile_args_struct" where it calls the “buffer_struct_argument" or "uniform_struct_argument"

Ok I see that you added it in this PR. Interesting that #4205 didn't have it but nothing was crashing though. 

So the only thing is to please move "compile_args_struct" out of `run()` and into `visit(ArgLoadStmt)` and I would be a happy, happy man.

Reply:
Let's leave it in `run` rn, afterwards we will have a prepass that scans all the required features / structures, and then we can clean up all the initialization code (instead of the one pass approach we are using rn). If we want to move it to `visit(ArgLoadStmt)` rn we will need an if statement to avoid generating multiple parts.

Reply:
And an if statement is bad because?

Reply:
It is not an elegant solution either. Proper fix will happen when we have a prepass featuer scan.

Reply:
its not that ugly either. we know that a visit(ArgLoad) exists iff you need the arg structs. so please move it :)

Reply:
No I don't think that's the right way. We shouldn't scatter global section initialization and type definitions across all the statements. We will add a prepass soon once the Metal random seed PR is merged.